distance-method/distance_method.py

```diff
@@ -0,0 +1,981 @@
+        if stderr and warnings:
+            print stderr
+
+
+def normalize_distances(phylip_fp,
+                        full_distance_matrix,
+                        num_species,
+                        full_distance_matrix_offset,
+                        species_set_dict,
+                        gene_bitvector_map,
+                        debug=False):
+    """ This function parses the output file of PHYLIP's protdist function
+    containing the distance matrix, Z-score normalizes the set of pairwise
+    distances between the gene in a species and all other species and stores
+    the results in a separate array.
+
+    Each normalized distance matrix is then sorted by species name and added
+    to the complete array storing distance matrices for all genes. In
+    addition, a list of missing species (species which did not include a
+    certain gene) is also maintained and used for setting nan's in array cells
+    which represent those species.
+
+    Below is an example of a parsed distance matrix
+    for 3 genes and 3 species:
+        0         1         2        (genes)
+    0_0 nan       nan       nan
+    0_1 0.53099   0.878855  0.83673
+    0_2 0.642856  1.083039  1.083039
+    1_0 0.300297  0.300297  0.702003
+    1_1 nan       nan       nan
+    1_2 0.399722  0.379156  0.356543
+    2_0 0.53099   0.53099   0.83673
+    2_1 0.399722  0.399722  0.356543
+    2_2 nan       nan       nan
+
+    (species pairs)
+
+    Example of Z-score normalized distance matrix from
+    above:
+        0            1           2          (genes)
+    0_0 nan          nan         nan
+    0_1 -1.40548346  0.83861735  0.56686611
+    0_2 -1.41421356  0.70710678  0.70710678
+    1_0 -0.70710678 -0.70710678  1.41421356
+    1_1 nan          nan         nan
+    1_2 1.20493966   0.03869341 -1.24363308
+    2_0 -0.70710678 -0.70710678  1.41421356
+    2_1 0.70710678   0.70710678 -1.41421356
+    2_2 nan          nan         nan
+
+    (species pairs)
+
+    Parameters
+    ----------
+    phylip_fp: string
+        filepath to distance matrix output by PHYLIP's protdist function
+    full_distance_matrix: dictionary
+        complete distance matrix for pairwise alignments between all species
+        for every gene
+    num_species: integer
+        number of species in the reference database
+    full_distance_matrix_offset: integer
+        the index offset for elements in full_distance_matrix where to write
+        the next array
+    species_set_dict: dictionary
+        dictionary containing the binary indicator vectors as keys and the
+        number of genes with identical species set represented by the binary
+        vectors as values
+    gene_bitvector_map: list
+        list containing the binary indicator vector for each query gene
+    debug: boolean
+        if True, run function in debug mode
+    """
+    # assume a pairwise alignment exists for all species
+    missing_species = [str(x) for x in range(0, num_species)]
```

Comment:
I don't think so because I need this array created once for each gene (normalize_distances() is called within a loop that iterates over MSA results of each gene in the query genome).